Reference/Common-Pitfalls/index-v9.md

```diff
@@ -0,0 +1,426 @@
+* Importing lots of data into Umbraco content nodes that could be stored in a custom database table (i.e. it's not going to be edited).
+In some cases this might be ok but many times we've seen bulk imports occur on an hourly/daily schedule which is generally unnecessary.
+
+## Processing during startup
+
+Umbraco allows you to run some initialization code during startup by using `UmbracoApplicationStartingNotification` or `UmbracoApplicationMainDomAcquiredNotification`, however great
+care should be used to ensure that you are not slowing down application startup. You should be especially careful
+as a Package developer that you are not slowing down application startup since your package may end up being used for
+thousands of websites.
+
+In many cases, [initialization code can be done lazily instead of eagerly](https://msdn.microsoft.com/en-us/library/dd997286(v=vs.110).aspx).
+Instead of initialization everything you need as soon as the application starts you could execute your initialization code only when it is required.
+This can be achieved in various ways such as:
+
+* Using [`Lazy<T>`](https://msdn.microsoft.com/en-us/library/dd642331(v=vs.110).aspx) and put the initialization logic in its callback
+* Using [`LazyInitializer`](https://msdn.microsoft.com/en-us/library/system.threading.lazyinitializer%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396)
+* Putting logic in a property getter with a lock and setting a flag that it's processed
+* Putting logic in a method with a lock and setting a flag that it's processed
+* (there's plenty of ways)
+
+Even more important is that you ensure that the initialization logic only executes one time for the lifetime of the
+application even when your app domain is restarted. If your initialization logic creates a database table or something similar to that, where it should only be executed one time only. Then you should set a persistent flag (such as a file) to indicate to your own logic that the initialization code has already executed and doesn't need to be done again.
+
+## Rebuilding indexes
+
+Far too often we've seen code in people's solutions that rebuild the Examine indexes
+(we've even seen this done on every request!). Rebuilding indexes can cause severe
+performance penalties and is not a recommended practice. Umbraco's and Examine's index management, index stability and
+synchronization of the data in the index gets better with every release. You should always ensure you are running the latest
+Umbraco and Examine versions if you are having trouble with your index data becoming out of sync with your Umbraco data.
+
+The primary reasons your data will become out of sync are:
+
+* Old version of Umbraco
+* Rebuilding indexes and restarting your app domain at the same time (try to avoid this scenario!)
+
+It is not recommended to rebuild your indexes unless you absolutely need to and if you need to do this often then it is
+advised to determine why and to try to resolve the underlying problem.
```

Comment:
```suggestion
## Rebuilding indexes

Far too often we've seen code in people's solutions that rebuild the Examine indexes
(we've even seen this done on every request!). Rebuilding indexes can cause severe
performance penalties and is not a recommended practice. Umbraco's and Examine's index management, index stability, and
synchronization of the data in the index get better with every release. You should always ensure you are running the latest
Umbraco and Examine versions if you are having trouble with your index data becoming out of sync with your Umbraco data.

The primary reasons your data will become out of sync are:

* Old version of Umbraco
* Rebuilding indexes and restarting your app domain at the same time (try to avoid this scenario!)

It is not recommended to rebuild your indexes unless you absolutely need to and if you need to do this often then it is
advised to determine why and to try to resolve the underlying problem.
```